ribbon-loadbalancer/src/main/java/com/netflix/loadbalancer/AbstractDynamicServerListLoadBalancer.java

```diff
@@ -0,0 +1,136 @@
+import com.netflix.servo.annotations.DataSourceType;
+import com.netflix.servo.annotations.Monitor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+public abstract class AbstractDynamicServerListLoadBalancer<T extends Server> extends BaseLoadBalancer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractDynamicServerListLoadBalancer.class);
+
+    // to keep track of modification of server lists
+    protected AtomicBoolean serverListUpdateInProgress = new AtomicBoolean(false);
+
+    protected volatile ServerList<T> serverListImpl;
+
+    protected volatile ServerListFilter<T> filter;
+
+    protected AtomicLong lastUpdated = new AtomicLong(System.currentTimeMillis());
+
+
+    protected AbstractDynamicServerListLoadBalancer() {
+        super();
+    }
+
+    protected AbstractDynamicServerListLoadBalancer(IClientConfig clientConfig, IRule rule, IPing ping) {
+        super(clientConfig, rule, ping);
+    }
+
+    @Override
+    public void setServersList(List lsrv) {
+        super.setServersList(lsrv);
+        List<T> serverList = (List<T>) lsrv;
+        Map<String, List<Server>> serversInZones = new HashMap<String, List<Server>>();
+        for (Server server : serverList) {
+            // make sure ServerStats is created to avoid creating them on hot
+            // path
+            getLoadBalancerStats().getSingleServerStat(server);
+            String zone = server.getZone();
+            if (zone != null) {
+                zone = zone.toLowerCase();
+                List<Server> servers = serversInZones.get(zone);
+                if (servers == null) {
+                    servers = new ArrayList<Server>();
+                    serversInZones.put(zone, servers);
+                }
+                servers.add(server);
+            }
+        }
+        setServerListForZones(serversInZones);
+    }
+
+    protected void setServerListForZones(Map<String, List<Server>> zoneServersMap) {
+        LOGGER.debug("Setting server list for zones: {}", zoneServersMap);
+        getLoadBalancerStats().updateZoneServerMapping(zoneServersMap);
+    }
+
+    public ServerList<T> getServerListImpl() {
+        return serverListImpl;
+    }
+
+    public void setServerListImpl(ServerList<T> niwsServerList) {
+        this.serverListImpl = niwsServerList;
+    }
+
+    public ServerListFilter<T> getFilter() {
+        return filter;
+    }
+
+    public void setFilter(ServerListFilter<T> filter) {
+        this.filter = filter;
+    }
+
+    private String getIdentifier() {
+        return this.getClientConfig().getClientName();
+    }
+
+    public abstract void stopServerListRefreshing();
+
+    @VisibleForTesting
+    public void updateListOfServers() {
```

Comment:
Might want to add a 'busy' AtomicBoolean to make sure this sit not called concurrently.  It's ok to drop the extra event.